### PR TITLE
Alerting: Fix rule editor issues with Azure Monitor data source

### LINKED ISSUE #41001: [ Azure Monitor ] [ alerting ] "failed to execute query A: no handler found for query type"
**What happened**:

When configuring Grafana alert rules, with Azure Monitor Datasource we had an error: 

`Failed to evaluate queries and expressions: failed to execute conditions: failed to execute query A: no handler found for query type`

**What you expected to happen**:

Alert rule to get created without errors.

**FIX**:

When investigating further we see the request payload is passing queryType as null, when we change it to Azure Monitor, it works fine. Screenshots with the default payload from Grafana Azure Monitor plugin and corrected payload follow. 

NOT WORKING:

image

WORKING:

image

**How to reproduce it (as minimally and precisely as possible)**:

Create an Alert rule with Azure Monitor data source. Please fix the payload.

**Anything else we need to know?**:

**Environment**:
- Grafana version: helm.sh/chart=grafana-6.16.4
- Grafana version: 8.1.0
- Data source type & version: Azure Monitor 0.3.0
- OS Grafana is installed on: Kube-prometheus-stack
- User OS & Browser: All
- Grafana plugins: Azure Monitor datasource
- Others:

### PR DESCRIPTION
**Pull request by domasx2**
Fixes two issues with Azure Monitor data source in rule editor page:
* Eval not working due to `queryType` not being set
* Visualization crashing due to incorrect default unit being used for threshold handle scale ( `undefined property 'ori'` error)
* Error spam while configuring the query, it would try to eval before query is fully configured. Updated `AlertQueryRunner` to use `DatasourceApi.filterQuery` to figure out if query is runnable

Note: 

https://user-images.githubusercontent.com/847684/140538052-7e435f43-751c-4ad0-a5df-b327a3a360cb.mp4

Fixes #41001

### CODE REVIEW

File: public/app/features/alerting/unified/components/rule-editor/VizWrapper.tsx [hunk 1]
```diff
@@ -98,13 +100,19 @@ const getStyles = (visHeight: number) => (theme: GrafanaTheme2) => ({
   `,
 });
 
-function defaultFieldConfig(thresholds: ThresholdsConfig): FieldConfigSource {
+function defaultUnit(data: PanelData): string | undefined {
+  return data.series[0]?.fields.find((field) => field.type === 'number')?.config.unit;
```
Comment (dprokop) [inline public/app/features/alerting/unified/components/rule-editor/VizWrapper.tsx:104]: Looks good

**Review by dprokop — COMMENTED** [review]
Don't think I have enough background to verify everything in this PR, but the viz/unit thing looks good 👍

File: public/app/features/alerting/unified/components/rule-editor/QueryRows.tsx [hunk 2]
```diff
@@ -136,6 +136,7 @@ export class QueryRows extends PureComponent<Props, State> {
         return {
           ...item,
           refId: query.refId,
+          queryType: item.model.queryType ?? '',
```
Comment (andresmgot) [inline public/app/features/alerting/unified/components/rule-editor/QueryRows.tsx:139]: would it make sense to simply "copy" the `model` in the root object? Something like:

```
return {
  ...item,
  ...item.model,
  refId: query.refId
```

Not sure if in the future another property apart from the `queryType` will be needed.

**Review by andresmgot — APPROVED** [review]
Thanks for fixing this! Not very familiar with this code but approving just in case.

File: public/app/features/alerting/unified/state/AlertingQueryRunner.ts [hunk 3]
```diff
@@ -41,12 +42,24 @@ export class AlertingQueryRunner {
     return this.subject.asObservable();
   }
 
-  run(queries: AlertQuery[]) {
+  async run(queries: AlertQuery[]) {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/state/AlertingQueryRunner.ts:45]: I think we should check every invocation of `AlertingQueryRunner.run()` to make sure that when `dataSourceSrv.get` throws we are not crashing the interface or make sure we have a `try..catch` mechanism here.

**Review by gillesdemey — CHANGES_REQUESTED** [review]

File: public/app/features/alerting/unified/state/AlertingQueryRunner.ts [hunk 4]
```diff
@@ -41,12 +42,24 @@ export class AlertingQueryRunner {
...
     if (queries.length === 0) {
       const empty = initialState(queries, LoadingState.Done);
       return this.subject.next(empty);
     }
 
+    // do not execute if one more of the queries are not runnable,
+    // for example not completely configured
+    for (const query of queries) {
+      if (query.datasourceUid !== ExpressionDatasourceUID) {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/state/AlertingQueryRunner.ts:54]: I think we can use `!isExpressionQuery` here as that will also check for `__expr__` types?

**Review by gillesdemey — APPROVED** [review]

**Pull request merged**